fv/wireguard_test.go

```diff
@@ -0,0 +1,444 @@
+					return wgPubKey
+				}, "5s", "100ms").Should(Equal(wgPubKeyOrig))
+			}
+		})
+
+		It("the Wireguard device should be configurable", func() {
+			// The Wireguard device should appear with default MTU, etc.
+			for _, felix := range felixes {
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "-d", "link", "show", wireguardInterfaceNameDefault)
+					return out
+				}, "10s", "100ms").Should(ContainSubstring(fmt.Sprintf("mtu %d", wireguardMTUDefault)))
+			}
+
+			// Change WireGuard configuration.
+			ifaceName := "wg0"
+			mtu := 1400
+			rule := 100
+			port := 28150
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			fc, err := client.FelixConfigurations().Get(ctx, "default", options.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			fc.Spec.WireguardInterfaceName = ifaceName
+			fc.Spec.WireguardMTU = &mtu
+			fc.Spec.WireguardListeningPort = &port
+			fc.Spec.WireguardRoutingRulePriority = &rule
+			_, err = client.FelixConfigurations().Update(ctx, fc, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			// Expect the settings to be changed on the device.
+			for _, felix := range felixes {
+				/* TODO: Enable after the fix.
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "-d", "link", "show", wireguardInterfaceNameDefault)
+					fmt.Println("link:", out)
+					return out
+				}, "10s", "100ms").Should(BeEmpty())
+				*/
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "-d", "link", "show", ifaceName)
+					fmt.Println("link:", out)
+					return out
+				}, "10s", "100ms").Should(ContainSubstring(fmt.Sprintf("mtu %d", mtu)))
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("/wg")
+					fmt.Println("wg:", out)
+					return out
+				}, "10s", "100ms").Should(ContainSubstring(fmt.Sprintf("listening port: %d", port)))
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "rule", "show", "pref", fmt.Sprintf("%d", rule))
+					fmt.Println("rule:", out)
+					return out
+				}, "10s", "1s").ShouldNot(BeEmpty())
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "rule", "show", "pref", wireguardRoutingRulePriorityDefault)
+					fmt.Println("rule:", out)
+					return out
+				}, "10s", "1s").Should(BeEmpty())
+			}
+		})
+	})
+
+	Context("with Wireguard disabled", func() {
+		BeforeEach(func() {
+			felixPID := felixes[0].GetFelixPID()
+
+			disableWireguard(client)
+
+			// Wait for felix to restart.
+			Eventually(felixes[0].GetFelixPID, "5s", "100ms").ShouldNot(Equal(felixPID))
+		})
+
+		It("the Wireguard device shouldn't exist", func() {
+			for _, felix := range felixes {
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "-d", "link", "show", wireguardInterfaceNameDefault)
+					return out
+				}, "10s", "100ms").Should(BeEmpty())
+			}
+		})
+
+		It("the Wireguard rule-entry/route-table shouldn't list wireguard rule", func() {
```

Comment:
I guess this is an example where an FV check is a good idea since it's hard to prove a negative otherwise.

Reply:
Suggest you make a common assertion function between the positive "rule exists" and the negative "rule doesn't exist" case.  Then you can do `Eventually(wireguardRuleExists).Should(BeTrue(), "wireguard rule was missing")` and  `Eventually(wireguardRuleExists).Should(BeFalse(), "wireguard rule was unexpectedly present")`

The presence of the positive version bolsters the negated check: if the code changes so the positive version fails to detect the rule, the FV will fail.